MerkleTreeConstructor.java

```diff
@@ -0,0 +1,99 @@
+package com.karolismed.simple_blockchain.blockchain;
+
+import com.karolismed.simple_blockchain.blockchain.model.transaction.Transaction;
+import com.karolismed.simple_blockchain.hashing.HashingService;
+import lombok.NonNull;
+import org.bitcoinj.core.Sha256Hash;
+import org.bitcoinj.core.Utils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.stream.Collectors;
+
+public class MerkleTreeConstructor {
+
+    private HashingService hashingService;
+
+    public MerkleTreeConstructor() {
+        this.hashingService = new HashingService();
+    }
+
+    public String getMerkleTreeRoot(@NonNull List<Transaction> transactions) {
+        return getMerkleTreeRootFromIds(
+            transactions.stream().map(Transaction::getTxId).collect(Collectors.toList())
+        );
+    }
+
+    public String getMerkleTreeRootFromIds(@NonNull List<String> transactionHashes) {
+        List<Sha256Hash> reversedHashes = transactionHashes.stream()
+            .map(hash -> Sha256Hash.wrapReversed(Utils.HEX.decode(hash.toLowerCase())))
+            .collect(Collectors.toList());
+        return Utils.HEX.encode(computeMerkleTreeRoot(reversedHashes).getReversedBytes());
+    }
+
+    private Sha256Hash computeMerkleTreeRoot(@NonNull List<Sha256Hash> merkle) {
+        // Stop if hash list is empty or contains one element
+        if (merkle.isEmpty()) {
+            return Sha256Hash.ZERO_HASH;
+        }
+        else if (merkle.size() == 1) {
+            return merkle.get(0);
+        }
+
+
+        // While there is more than 1 hash in the list, keep looping...
+        while (merkle.size() > 1) {
+            // If number of hashes is odd, duplicate last hash in the list.
+            if (merkle.size() % 2 != 0) {
+                merkle.add(merkle.get(merkle.size() - 1));
+            }
+            assert merkle.size() % 2 == 0;
+
+            // New hash list.
```

Comment:
Once again, excessive comment indicating a statement that is already clear enough to be interpreted by itself.

Reply:
Addressed